crates/typst-ide/src/definition.rs

```diff
@@ -0,0 +1,243 @@
+            NamedItem::Import(name, span, value) => {
+                Some(Definition::item(name, Span::detached(), span, value.cloned()))
+            }
+        }
+    });
+    let src = src.or_else(|| {
+        let in_math = matches!(
+            leaf.parent_kind(),
+            Some(SyntaxKind::Equation)
+                | Some(SyntaxKind::Math)
+                | Some(SyntaxKind::MathFrac)
+                | Some(SyntaxKind::MathAttach)
+        );
+        let library = world.library();
+
+        let scope = if in_math { library.math.scope() } else { library.global.scope() };
+        for (item_name, value) in scope.iter() {
+            if item_name == name {
+                return Some(Definition::item(
+                    name,
+                    Span::detached(),
+                    Span::detached(),
+                    Some(value.clone()),
+                ));
+            }
+        }
+
+        None
+    });
+
+    if paths.is_empty() {
+        return src;
+    }
+
+    None
+}
+
+/// A definition of some item.
+#[derive(Debug, Clone)]
+pub struct Definition {
+    /// The name of the definition.
+    pub name: EcoString,
+    /// The kind of the definition.
+    pub kind: DefinitionKind,
+    /// A possible instance of the definition.
+    pub value: Option<Value>,
+    /// The source span of the entire definition.
+    pub span: Span,
+    /// The span of the name of the definition.
+    pub name_span: Span,
+}
+
+impl Definition {
+    fn with_kind(self, kind: DefinitionKind) -> Self {
+        Self { kind, ..self }
+    }
+
+    fn item(name: &EcoString, span: Span, name_span: Span, value: Option<Value>) -> Self {
+        let kind = value
+            .as_ref()
+            .and_then(|e| {
+                matches!(e, Value::Func(..)).then_some(DefinitionKind::Function)
+            })
+            .unwrap_or(DefinitionKind::Variable);
+
+        Self { name: name.clone(), kind, value, span, name_span }
+    }
+
+    fn module(module: &Module) -> Self {
+        Definition {
+            name: EcoString::new(),
+            kind: match module.file_id() {
+                Some(file_id) => DefinitionKind::ModulePath(file_id),
+                None => DefinitionKind::Module(module.clone()),
+            },
+            value: Some(Value::Module(module.clone())),
+            span: Span::detached(),
+            name_span: Span::detached(),
+        }
+    }
+}
+
+/// A kind of item that is definition.
+#[derive(Debug, Clone, PartialEq, Hash)]
+pub enum DefinitionKind {
+    /// ```plain
+    /// let foo;
+    ///     ^^^
+    /// ```
+    Variable,
+    /// ```plain
+    /// let foo(it) = it;
+    ///     ^^^
+    /// ```
+    Function,
+    /// ```plain
+    /// import ("fo" + "o.typ")
+    ///          ^^^^^^^^
+    /// ```
+    ///
+    /// IDE will always resolve a path instead of a module whenever possible.
+    /// This allows resolving a module containing errors.
+    ModulePath(FileId),
```

Comment:
I understand the desire to be resilient, but since the code currently only generates a `ModulePath` from an already known `Module` value it feels unnecessary for the time being.